src/rust/src/wallet_scanner.rs

```diff
@@ -0,0 +1,470 @@
+                consensus::NetworkUpgrade::Sapling => *sapling,
+                consensus::NetworkUpgrade::Blossom => *blossom,
+                consensus::NetworkUpgrade::Heartwood => *heartwood,
+                consensus::NetworkUpgrade::Canopy => *canopy,
+                consensus::NetworkUpgrade::Nu5 => *nu5,
+            },
+        }
+    }
+
+    fn coin_type(&self) -> u32 {
+        match self {
+            Self::Consensus(params) => params.coin_type(),
+            Self::RegTest { .. } => constants::regtest::COIN_TYPE,
+        }
+    }
+
+    fn hrp_sapling_extended_spending_key(&self) -> &str {
+        match self {
+            Self::Consensus(params) => params.hrp_sapling_extended_spending_key(),
+            Self::RegTest { .. } => constants::regtest::HRP_SAPLING_EXTENDED_SPENDING_KEY,
+        }
+    }
+
+    fn hrp_sapling_extended_full_viewing_key(&self) -> &str {
+        match self {
+            Self::Consensus(params) => params.hrp_sapling_extended_full_viewing_key(),
+            Self::RegTest { .. } => constants::regtest::HRP_SAPLING_EXTENDED_FULL_VIEWING_KEY,
+        }
+    }
+
+    fn hrp_sapling_payment_address(&self) -> &str {
+        match self {
+            Self::Consensus(params) => params.hrp_sapling_payment_address(),
+            Self::RegTest { .. } => constants::regtest::HRP_SAPLING_PAYMENT_ADDRESS,
+        }
+    }
+
+    fn b58_pubkey_address_prefix(&self) -> [u8; 2] {
+        match self {
+            Self::Consensus(params) => params.b58_pubkey_address_prefix(),
+            Self::RegTest { .. } => constants::regtest::B58_PUBKEY_ADDRESS_PREFIX,
+        }
+    }
+
+    fn b58_script_address_prefix(&self) -> [u8; 2] {
+        match self {
+            Self::Consensus(params) => params.b58_script_address_prefix(),
+            Self::RegTest { .. } => constants::regtest::B58_SCRIPT_ADDRESS_PREFIX,
+        }
+    }
+}
+
+/// A decrypted note.
+struct DecryptedNote<D: Domain> {
+    /// The incoming viewing key used to decrypt the note.
+    ivk: D::IncomingViewingKey,
+    /// The recipient of the note.
+    recipient: D::Recipient,
+    /// The note!
+    note: D::Note,
+    /// The memo sent with the note.
+    memo: D::Memo,
+}
+
+/// A value correlated with an output index.
+struct OutputIndex<V> {
+    /// The index of the output within the corresponding shielded bundle.
+    output_index: usize,
+    /// The value for the output index.
+    value: V,
+}
+
+/// Extension trait for things `zcash_note_encryption` doesn't provide that the `zcashd`
+/// wallet needs.
+trait DomainExt: BatchDomain {
+    /// Figures out which incoming viewing key matches the given recipient.
+    ///
+    /// This is linear in the size of `ivks`, but importantly this is only run when an
+    /// output is successfully decrypted, and therefore is not linear in the number of
+    /// on-chain outputs.
+    fn ivk_for_recipient(
+        ivks: &[Self::IncomingViewingKey],
```

Comment:
The implementation doesn't depend on this being a slice, it could be an iterable.

Reply:
I've deleted this as it was actually unnecessary.